### PR TITLE
Add the ability to view and list of leases metadata

### PR DESCRIPTION
**Pull request by chrishoffman**
This adds the ability to view and list lease metadata.  There are currently no methods to view lease metadata or to list leases by a given prefix.  Since there are methods that exist to revoke individual or a prefix of leases and renew individual leases, it makes sense to have endpoints that allow you to view this data as well.  The data that will be returned will only be metadata about the lease and will not return any internal lease data.

### CODE REVIEW

File: vault/logical_system.go [hunk 1]
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+		prefix = prefix + "/"
+	}
+
+	keys, err := b.Core.expiration.idView.List(prefix)
+	if err != nil {
+		b.Backend.Logger().Error("sys: error listing leases", "prefix", prefix, "error", err)
+		return handleError(err)
+	}
+
+	return &logical.Response{
```
Comment (jefferai) [inline vault/logical_system.go]: Check `logical.ListResponse`, which does this for you!

File: vault/logical_system.go [hunk 2]
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
 	return nil, nil
 }
 
+// handleLeasse is use to view the metadata for a given LeaseID
+func (b *SystemBackend) handleLease(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	leaseID := data.Get("lease_id").(string)
+
```
Comment (vishalnayak) [inline vault/logical_system.go]: Can we have a check for `leaseID == ""` here?

File: vault/logical_system.go [hunk 3]
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+	}
+	if leaseTimes == nil {
+		return logical.ErrorResponse("invalid lease"), logical.ErrInvalidRequest
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"lease_id":          leaseID,
+			"issue_time":        leaseTimes.IssueTime,
+			"expire_time":       nil,
```
Comment (vishalnayak) [inline vault/logical_system.go]: Is there a reason for setting these 2 fields explicitly to `nil`?

File: vault/logical_system_test.go [hunk 4]
```diff
@@ -315,6 +317,179 @@ func TestSystemBackend_remount_system(t *testing.T) {
...
+	resp, err := core.HandleRequest(req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if resp != nil {
+		t.Fatalf("bad: %#v", resp)
+	}
+
+	// Read a key with a LeaseID
+	req = logical.TestRequest(t, logical.ReadOperation, "secret/foo")
```
Comment (vishalnayak) [inline vault/logical_system_test.go]: Generic secret backend does not attach a lease to secrets. I'm not sure if this test would work.

File: vault/logical_system_test.go [hunk 5]
```diff
@@ -315,6 +317,179 @@ func TestSystemBackend_remount_system(t *testing.T) {
...
```
Comment (vishalnayak) [inline vault/logical_system_test.go]: Same with this test as well.

File: vault/logical_system_test.go (hunk 4 above)
Comment (chrishoffman) [inline vault/logical_system_test.go]: It does and I was surprised too and this how all the renew and revoke functions are written.  I think it might be something in the test harness.  Otherwise it isn't straightforward creating a non-token lease without some backing dynamic backend.  Maybe `pki` with `generate_lease` set `true`.

File: vault/logical_system.go [hunk 6]
```diff
@@ -299,6 +300,42 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
...
+				Callbacks: map[logical.Operation]framework.OperationFunc{
+					logical.UpdateOperation: b.handleLease,
+				},
+
+				HelpSynopsis:    strings.TrimSpace(sysHelp["lease"][0]),
+				HelpDescription: strings.TrimSpace(sysHelp["lease"][1]),
+			},
+
+			&framework.Path{
+				Pattern: "lease" + framework.OptionalParamRegex("prefix"),
```
Comment (briankassouf) [inline vault/logical_system.go]: Instead of making this optional what if the two framework.Paths were merged with both a read op and a list op?

File: vault/logical_system.go (hunk 3 above)
Comment (chrishoffman) [inline vault/logical_system.go]: This was really just to serialize to `null` in the serializer.  I guess omitting would probably be fine.

File: vault/logical_system.go [hunk 7]
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+		resp.Data["renewable"] = leaseTimes.Secret.Renewable
+		resp.Data["ttl"] = leaseTimes.Secret.TTL
+	}
+
+	return resp, nil
+}
+
+func (b *SystemBackend) handleLeaseList(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	prefix := data.Get("prefix").(string)
```
Comment (briankassouf) [inline vault/logical_system.go]: Can you add a check to make sure prefix is not `""`

File: vault/logical_system.go (hunk 2 above)
Comment (chrishoffman) [inline vault/logical_system.go]: Will do.

File: vault/logical_system.go (hunk 1 above)
Comment (chrishoffman) [inline vault/logical_system.go]: Magic :).

File: vault/logical_system.go [hunk 8]
```diff
@@ -63,6 +63,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 				"replication/reindex",
 				"rotate",
 				"config/auditing/*",
+				"lease*",
```
Comment (briankassouf) [inline vault/logical_system.go]: Also you'll need to update the root paths test after changing this list

File: vault/logical_system.go [hunk 9]
```diff
@@ -299,6 +300,42 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
...
+
+				Fields: map[string]*framework.FieldSchema{
+					"lease_id": &framework.FieldSchema{
+						Type:        framework.TypeString,
+						Description: strings.TrimSpace(sysHelp["lease_id"][0]),
+					},
+				},
+
+				Callbacks: map[logical.Operation]framework.OperationFunc{
+					logical.UpdateOperation: b.handleLease,
```
Comment (chrishoffman) [inline vault/logical_system.go]: Since `lease_id` is sensitive, we determined it made sense to pass `lease_id` as a parameter with an update operation.  This is currently the preferred method to use in `/sys/renew` and `/sys/revoke`.

File: vault/logical_system.go (hunk 7 above)
Comment (chrishoffman) [inline vault/logical_system.go]: Empty prefix is valid and will list the top level keys.

File: vault/logical_system.go (hunk 6 above)
Comment (chrishoffman) [inline vault/logical_system.go]: My only concern here was some arbitrary path would not be an "unsupported operation" on update operations because of the optional `prefix`.

File: vault/logical_system.go (hunk 8 above)
Comment (chrishoffman) [inline vault/logical_system.go]: Thanks for the head up.  I wasn't aware of that test.

File: vault/logical_system_test.go (hunk 4 above)
Comment (jefferai) [inline vault/logical_system_test.go]: When we removed leases from the generic backend in the normal case we still kept the ability to have it issue leases specifically to make testing easier!

File: vault/logical_system.go [hunk 10]
```diff
@@ -347,7 +371,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 			},
 
 			&framework.Path{
-				Pattern: "revoke-force/(?P<prefix>.+)",
+				Pattern: "(lease/)?revoke-force" + framework.OptionalParamRegex("prefix"),
```
Comment (vishalnayak) [inline vault/logical_system.go]: Before this change, I think the `prefix` parameter was mandatory and would always be set. If it was not supplied, parsing the path pattern would error out. Now that it is an optional param, missing `prefix` I think will not error out. We might want to check explicitly that the required parameters (but optionally set in the URL) are in fact set. This comment applies to all the similar changes below.

Comment (chrishoffman) [inline vault/logical_system.go]: You are right.  I got a little ambitious in my refactoring.  I'll revert out these changes.

File: vault/logical_system.go [hunk 11]
```diff
@@ -55,14 +55,16 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
...
-				"revoke-prefix/*",
 				"audit",
 				"audit/*",
 				"raw/*",
 				"replication/primary/secondary-token",
 				"replication/reindex",
 				"rotate",
 				"config/auditing/*",
+				"lease/lookup*",
+				"lease/revoke-prefix/*",
```
Comment (vishalnayak) [inline vault/logical_system.go]: Can we also add `revoke-force/*` and `lease/revoke-force/*` to this list?

File: vault/logical_system.go [hunk 12]
```diff
@@ -1274,6 +1300,56 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"id":            leaseID,
+			"creation_time": leaseTimes.IssueTime,
+			"renewable":     leaseTimes.renewable(),
+		},
+	}
+	if !leaseTimes.LastRenewalTime.IsZero() {
```
Comment (vishalnayak) [inline vault/logical_system.go]: Sorry for circling around on this. On a second thought, the idea of having fields with default values (like you were doing earlier) does seem like a sensible thing to do too. What might be of concern is that this API will not have deterministic number of response fields. That is okay I guess but I am not sure if that is a problem. Bringing this up so we can all agree on this and move forward.

File: vault/logical_system.go [hunk 13]
```diff
@@ -2429,4 +2505,19 @@ This path responds to the following HTTP methods.
 		"Lists the headers configured to be audited.",
 		`Returns a list of headers that have been configured to be audited.`,
 	},
+
+	"lease": {
+		``,
```
Comment (vishalnayak) [inline vault/logical_system.go]: Reminder to fill these out.

File: vault/logical_system.go [hunk 14]
```diff
@@ -55,14 +55,16 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
...
 				"remount",
-				"revoke-prefix/*",
 				"audit",
 				"audit/*",
 				"raw/*",
 				"replication/primary/secondary-token",
 				"replication/reindex",
 				"rotate",
 				"config/auditing/*",
+				"lease/lookup*",
```
Comment (jefferai) [inline vault/logical_system.go]: Can this be `leases` not `lease`? It's a collection so plural makes sense. (As opposed to a singular system like `replication`, e.g. if it was `lease-management`, but that's too wordy.)

File: vault/expiration.go [hunk 15]
```diff
@@ -841,11 +841,18 @@ type leaseEntry struct {
...
 
-func (le *leaseEntry) renewable() error {
+func (le *leaseEntry) renewable() bool {
+	if err := le.renewableErr(); err == nil {
+		return true
+	}
+	return false
+}
+
+func (le *leaseEntry) renewableErr() error {
```
Comment (jefferai) [inline vault/expiration.go]: What is the point of doing this?

File: vault/logical_system.go [hunk 16]
```diff
@@ -1274,6 +1300,56 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+		return handleError(err)
+	}
+	if leaseTimes == nil {
+		return logical.ErrorResponse("invalid lease"), logical.ErrInvalidRequest
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"id":            leaseID,
+			"creation_time": leaseTimes.IssueTime,
```
Comment (jefferai) [inline vault/logical_system.go]: This has a different type than `creation_time` in the token store; there, you set this to `issue_time`.

Comment (jefferai) [inline vault/logical_system.go]: There's a type discrepancy with `last_renewal_time` too.

File: vault/expiration.go (hunk 15 above)
Comment (chrishoffman) [inline vault/expiration.go]: I thought this was a bit confusing and was just consolidating the logic for dealing with the error, which the expiration manager uses, and returning the value of renewable.

File: vault/logical_system.go (hunk 14 above)
Comment (chrishoffman) [inline vault/logical_system.go]: I debated this but wanted to match the other paths that had similar functions, like `token` and `policy`.  I am not strongly either way and was just trying for consistency.

File: vault/logical_system.go (hunk 11 above)
Comment (chrishoffman) [inline vault/logical_system.go]: It is probably the right thing to do but this would be a change in behavior.  I just want to make sure that what we want to do.  I will note it in the changelog if we decide to go that route.

Comment (vishalnayak) [inline vault/logical_system.go]: If `revoke-prefix` is a privileged operation, `revoke-force` must be I suppose. If we decide to do this, we should document it in CL as you said, in the docs and as a breaking behavior in the upgrade notes.

File: vault/expiration.go (hunk 15 above)
Comment (jefferai) [inline vault/expiration.go]: I'm concerned about people defaulting to use this function when in fact the exact error returned can be useful, e.g. in a debugging/logs context.

File: vault/logical_system.go (hunk 14 above)
Comment (jefferai) [inline vault/logical_system.go]: The difference with `policy` is that it's supposed to be you acting on a single one, e.g. PUT (write) a policy, GET a policy, and so on. If there were a collection of policy-related APIs under a prefix, having that prefix be `policies` instead of `policy` would be the right move IMHO.

Comment (jefferai) [inline vault/logical_system.go]: Also, lease lookup should definitely not be a root path.

File: vault/logical_system.go (hunk 11 above)
Comment (jefferai) [inline vault/logical_system.go]: Generally we reserve root paths for functions that can be super destructive or super sensitive from a security standpoint. `revoke-force` does fit that bill and probably should be privileged, but I wonder if we should reserve that change for 0.8; people will be more likely to expect breaking changes for that release and I don't think this is super high priority.

**Comment by chrishoffman** [comment]
This should be ready for review again.  

Notable changes since last full review:
* Added documentation
* Nested "lease" related paths under `/sys/leases`.  Backwards compatibility is maintained.
* Token lookup and lease lookup fields merged

I ended up moving a lot of documentation around.  Should we provide landing pages for the old docs locations that point to the new locations?

File: vault/logical_system.go [hunk 17]
```diff
@@ -325,7 +363,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 			},
 
 			&framework.Path{
-				Pattern: "revoke" + framework.OptionalParamRegex("url_lease_id"),
+				Pattern: "(leases/)?revoke" + framework.OptionalParamRegex("url_lease_id"),
```
Comment (briankassouf) [inline vault/logical_system.go]: 👍

File: vault/logical_system.go [hunk 18]
```diff
@@ -1274,6 +1314,61 @@ func (b *SystemBackend) handleTuneWriteCommon(
 	return nil, nil
 }
 
+// handleLease is use to view the metadata for a given LeaseID
```
Comment (vishalnayak) [inline vault/logical_system.go]: s/handleLease/handleLeaseLookup

File: vault/logical_system.go [hunk 19]
```diff
@@ -1274,6 +1314,61 @@ func (b *SystemBackend) handleTuneWriteCommon(
...
+
+	leaseTimes, err := b.Core.expiration.FetchLeaseTimes(leaseID)
+	if err != nil {
+		b.Backend.Logger().Error("sys: error retrieving lease", "lease_id", leaseID, "error", err)
+		return handleError(err)
+	}
+	if leaseTimes == nil {
+		return logical.ErrorResponse("invalid lease"), logical.ErrInvalidRequest
+	}
+
```
Comment (vishalnayak) [inline vault/logical_system.go]: Can we add a comment here explaining why we are setting values to nil beforehand. This will avoid people removing it in future.

File: vault/logical_system.go [hunk 20]
```diff
@@ -299,7 +302,43 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
...
+				Callbacks: map[logical.Operation]framework.OperationFunc{
+					logical.ListOperation: b.handleLeaseLookupList,
+				},
+
+				HelpSynopsis:    strings.TrimSpace(sysHelp["leases"][0]),
+				HelpDescription: strings.TrimSpace(sysHelp["leases"][1]),
+			},
+
+			&framework.Path{
+				Pattern: "leases/lookup",
```
Comment (briankassouf) [inline vault/logical_system.go]: Do we want this to be a root protected path?

Comment (jefferai) [inline vault/logical_system.go]: @briankassouf Since `lease/lookup` doesn't return sensitive information (e.g. the contents of internal data) and you have to know the lease ID ahead of time I don't think it needs to be root. In fact, I am actually thinking we may want to add it to the `default` policy. _Listing_ on the other hand *does* divulge lease IDs which can be used to e.g. renew them, so that should be root protected.

**Comment by jefferai** [comment]
@vishalnayak it's not a breaking change; `sys/revoke-force` remains non-root (for now), but `sys/leases/revoke-force` is root. Since that's what documentation will focus on we'll get people used to using that, then eventually can change over the older path.
